### PR TITLE
Fix: align OSHMEM API with spec v1.2

### PR DESCRIPTION
**Pull request by igor-ivanov**
Port related PR https://github.com/open-mpi/ompi/pull/1149

:bot:label:bug
:bot:label:enhancement
:bot:milestone:v2.0.0
:bot:assign: @miked-mellanox

### CODE REVIEW

**Comment by mellanox-github** [comment]
Test PASSed.
See http://bgate.mellanox.com/jenkins/job/gh-ompi-release-pr/1068/ for details.

**Comment by mike-dubman** [comment]
:+1:

File: oshmem/include/pshmemx.h [hunk 1]
```diff
@@ -16,110 +16,194 @@
...
+OSHMEM_DECLSPEC  int p_num_pes(void);
+OSHMEM_DECLSPEC  int p_my_pe(void);
+
+OSHMEM_DECLSPEC  void* pshmalloc(size_t size);
+OSHMEM_DECLSPEC  void* pshmemalign(size_t align, size_t size);
+OSHMEM_DECLSPEC  void* pshrealloc(void *ptr, size_t size);
+OSHMEM_DECLSPEC  void pshfree(void* ptr);
+
+OSHMEM_DECLSPEC  void pshmem_char_put(char *target, const char *source, size_t len, int pe);
+OSHMEM_DECLSPEC  void pshmem_char_get(char *target, const char *source, size_t len, int pe);
```
Comment (jsquyres) [inline oshmem/include/pshmemx.h:35]: @igor-ivanov Do you want to use the deprecated pragmas here?  See mpi.h for an example, if you care.

Comment (hppritcha) [inline oshmem/include/pshmemx.h:35]: I think this would be helpful for users too.

**Comment by hppritcha** [comment]
@igor-ivanov are you intending to update the oshmem java bindings at some point as well to be more compatible with openshmem 1.2 spec?

**Comment by mike-dubman** [comment]
@hppritcha - java bindings for OSHMEM are not part of the OpenSHMEM SPEC.

Current OSHMEM/java author and last known maintainer (@osvegis) expressed no intent to support java part.
We can remove it if no interest.
We have interest in OSHMEM C and Fortan features only.

**Comment by jsquyres** [comment]
@miked-mellanox If you guys don't really want to maintain it, I kinda doubt anyone else is either (@hppritcha -- is LANL?).  If no one else is maintaining it, perhaps someone could file a future PR to remove the Java OSHMEM support (assuming no one is actually using it... do we know that for a fact?).

**Comment by hppritcha** [comment]
LANL is interested in maintaining/enhancing the MPI JAVA bindings but not the shmem ones.

File: oshmem/include/pshmemx.h (hunk 1 above)
Comment (igor-ivanov) [inline oshmem/include/pshmemx.h:35]: I believe that it is a good point but this capability requires changes that probably too heavy if follow right way.
We need to introduce new configure option as oshmem-interface-warning and make it consistent with mpi-interface-warning, duplicate mpi_portable_platform.h file to recogize a compiler and include it into shmem.h header. Define new or use OMPI_BUILDING inside shmem.h etc. I am not sure if we can include additional files inside shmem.h.
So I do not have clear point on this proposal.
@miked-mellanox what do you think?

Comment (jsquyres) [inline oshmem/include/pshmemx.h:35]: @igor-ivanov Ok, no prob.  Just wanted to make you aware that this kind of capability is available.  Seems like this PR should be merged in, and if you guys later want to add deprecation pragmas, that can be done in a future PR.  Sound good?

**Comment by jsquyres** [comment]
@hppritcha @miked-mellanox @igor-ivanov Ok, sounds like no one is interested in maintaining the Java OSHMEM bindings.  Would you guys mind removing them at some point before v2.0.0?

**Comment by jsquyres** [comment]
open-mpi/ompi#1149 has been merged into master; this is clear for v2.x.

**Comment by jsquyres** [comment]
@hppritcha I think this one is good to go.

**Comment by igor-ivanov** [comment]
@jsquyres I think I can update this PR with commit that removes Java as we agreed before merge. What do you think?

**Comment by mike-dubman** [comment]
@igor-ivanov - lets wait w remove on v1.10 and deal w/ non-production branches.

**Comment by igor-ivanov** [comment]
@miked-mellanox ok

**Comment by jsquyres** [comment]
+1 -- let's remove the java from master and v2.x, but if it's already in v1.10, then leave it there.

**Comment by hppritcha** [comment]
so do we want to include commits for removing oshmem java bindings with this PR?  Or have a subsequent PR which only has the bindings removal?

**Comment by jsquyres** [comment]
@hppritcha 6 of one, half dozen of the other, IMHO.  However Mellanox/@igor-ivanov wants to do it.  :-)

**Comment by hppritcha** [comment]
okay, let's wait.

File: oshmem/include/pshmemx.h (hunk 1 above)
Comment (igor-ivanov) [inline oshmem/include/pshmemx.h:35]: @jsquyres Personally I would like to provide a deprecation message to an user but using more easy solution. Unfortunately I do no see a simple way to do it following current openmpi scheme. I agree we can do it later.

**Comment by mellanox-github** [comment]
Test PASSed.
See http://bgate.mellanox.com/jenkins/job/gh-ompi-release-pr/1089/ for details.

**Comment by mike-dubman** [comment]
:+1:

**Comment by mellanox-github** [comment]
Test PASSed.
See http://bgate.mellanox.com/jenkins/job/gh-ompi-release-pr/1090/ for details.

**Comment by igor-ivanov** [comment]
@hppritcha this PR updated with JAVA remove commit ported from master: https://github.com/open-mpi/ompi/pull/1176

**Comment by hppritcha** [comment]
@jsquyres I think this is ready to merge.

**Comment by jsquyres** [comment]
Thanks @igor-ivanov

**Pull request merged**